### PR TITLE
program changed from ruby to python

### PR DESCRIPTION
**Pull request by shruti19**

### CODE REVIEW

File: dp/max_product_subarray.py [hunk 1]
```diff
@@ -19,3 +19,37 @@ def max_product(nums):
...
+        max_product_end *= arr[i]
+        if max_product_end < 0:
+            max_product_end = 1
+            max_start_i = i + 1
+
+        if product_so_far < max_product_end:
+            product_so_far = max_product_end
+            so_far_end_i = i
+            so_far_start_i = max_start_i
+    print "max_product_so_far: %s, %s"%(product_so_far, arr[so_far_start_i:so_far_end_i+1])
```
Comment (ankit167) [inline dp/max_product_subarray.py]: 1. missing whitespace around modulo operator
2. line too long (91 > 79 characters)

Comment (shruti19) [inline dp/max_product_subarray.py]: Sorry for the trouble @ankit167. I have pushed the correction.

File: dp/max_product_subarray.py [hunk 2]
```diff
@@ -19,3 +19,38 @@ def max_product(nums):
...
+        max_product_end *= arr[i]
+        if max_product_end < 0:
+            max_product_end = 1
+            max_start_i = i + 1
+
+        if product_so_far < max_product_end:
+            product_so_far = max_product_end
+            so_far_end_i = i
+            so_far_start_i = max_start_i
+    print "max_product_so_far: %s, %s" % (product_so_far,\
```
Comment (ankit167) [inline dp/max_product_subarray.py]: nit, '\' is not required insides parenthesis.

File: dp/max_product_subarray.py [hunk 3]
```diff
@@ -19,3 +19,38 @@ def max_product(nums):
...
+	arr[so_far_start_i:so_far_end_i + 1])
```
Comment (ankit167) [inline dp/max_product_subarray.py]: Please indent this line according to opening parenthesis (as per pep8 guidelines)

**Comment by ankit167** [comment]
The code fails for some of cases:
1. [-4, -3. -2. -1] (All integers are negative)
2. [-4, 0, 9, 8] (One or more integers are '0')
3. [-4, -3, -2, 1]

Please fix the logic to handle these cases, and other edge cases you can think of.

File: dp/max_product_subarray.py [hunk 4]
```diff
@@ -19,3 +19,38 @@ def max_product(nums):
...
+
+subarray_with_max_product([2,3,6,-1,-1,4,5])
+max_product_so_far: 36, [2, 3, 6]
+
+subarray_with_max_product([-2,-3,6,-1,-9,-5])
+max_product_so_far: 6, [6]
+'''
+
+def subarray_with_max_product(arr):
+    ''' arr is list of positive/negitive numbers '''
```
Comment (ankit167) [inline dp/max_product_subarray.py]: Typo- negative

**Pull request merged**